### PR TITLE
Adds ability to fork application via initialize_from

### PR DESCRIPTION
**Pull request by skrawcz**
This gets at #181, by enabling someone
to fork an application easily from prior state.

You don't need to manage this outside of Burr,
it can now be done with the existing persisters.

I have included various error checks, just need
to write some tests for them.

## Changes
- ApplicationBuilder to take in some more parameters
- updated conversational RAG notebook to show case the new feature

## How I tested this
 - locally

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: burr/core/application.py [hunk 1]
```diff
@@ -1395,6 +1401,9 @@ def initialize_from(
         :param resume_at_next_action: Whether to resume at the next action, or default to the ``default_entrypoint``
         :param default_state: The default state to use if it does not exist. This is a dictionary.
         :param default_entrypoint: The default entry point to use if it does not exist or you elect not to resume_at_next_action.
+        :param fork_from_app_id: The app ID to fork from. This is used to fork from a prior application run.
```
Comment (elijahbenizzy) [inline burr/core/application.py:1404]: Worth mentioning how these differ from `with_identifiers` -- this is a bit confusing

File: burr/core/application.py [hunk 2]
```diff
@@ -1403,10 +1412,18 @@ def initialize_from(
                 + "Cannot call initialize_from if you have already set state or an entrypoint! "
                 "You can either use the initializer *or* set the state and entrypoint manually."
             )
+        if not fork_from_app_id and (fork_from_partition_key or fork_from_sequence_id):
+            raise ValueError(
+                ERROR_MESSAGE
+                + "If you set fork_from_partition_key or fork_from_sequence_id, you must also set fork_from_app_id."
```
Comment (elijahbenizzy) [inline burr/core/application.py:1418]: Link to docs?

File: burr/core/application.py [hunk 3]
```diff
@@ -1441,22 +1458,47 @@ def _load_from_persister(self):
...
+                    f"{self.initializer.__class__.__name__} returned None while trying to fork from: "
+                    f"partition_key:{_partition_key}, app_id:{_app_id}, "
+                    f"sequence_id:{_sequence_id}. "
+                    "You explicitly requested to fork from a prior application run, but it does not exist. "
+                    "Defaulting to state defaults instead."
+                )
             # there was nothing to load -- use default state
             self.state = self.state.update(**self.default_state)
             self.sequence_id = None  # has to start at None
+            self.parent_id = None
```
Comment (elijahbenizzy) [inline burr/core/application.py:1490]: Comment as this is a placeholder

File: docs/concepts/state-persistence.rst [hunk 4]
```diff
@@ -38,11 +38,24 @@ This action takes in an initializer (an implementation of :py:class:`StateInitia
...
 
 Note that you cannot use this in conjunction with :py:meth:`with_state <burr.core.application.ApplicationBuilder.with_state>`
 or :py:meth:`with_entrypoint <burr.core.application.ApplicationBuilder.with_entrypoint>` -- these are mutually exclusive.
 Either you load from state or you start from scratch.
 
+Forking State
+_____________
+Note that you can also fork state from a previous application. This is useful if you want to start from a previous application's state,
+but don't want to add to it.The ``fork_from_app_id`` and ``fork_from_partition_key`` are used to identify the application to fork from, while
+``fork_from_sequence_id`` is used to identify the sequence_id to use. This is useful if you want to fork from a specific point in the application,
```
Comment (elijahbenizzy) [inline docs/concepts/state-persistence.rst:54]: Make more explicit the relationship between that/`with_identifiers` -- I think this will trip people up

**Review by elijahbenizzy — COMMENTED** [review]
Looks good, just some small comments on documentation

File: burr/core/application.py (hunk 3 above)
Comment (skrawcz) [reply burr/core/application.py:1490]: don't need to include it yet. removing

**Review by elijahbenizzy — APPROVED** [review]

**Pull request merged**